# Обработка исключений

> Ключевые слова
> `try`, `catch`, `throw`, `throws`, `finally`

## Пример использования блока исключений

```java

try {
  ...
} catch (Exception exception) {
  ...
} catch (RuntimeException exception) {
  ...
} finally {
  ...
}

```

**Заметки**
* catch будут перехватывать исключения в порядке объявления
* Применяя несколько операторов catch, важно помнить, что перехват исключений из подклассов должен следовать до перехвата исключений из суперклассов

## Иерархия исключений

![Alt text](img/9.error_hierarchy.png "Иерархия исключений")

 - **Exception** - ошибки прикладной программы. После возникновения штатное выполнение программы возможно. 
 - **Error** - ошибки уровня виртуальной машины. После возникновения штатное выполнение программы невозможно. 

## Throws

```java

void function testException() throws MyException
{
   ...
}

```

## Проверяеыме и непроверяемые исключения

![Alt text](img/9.checkend_and_unckecked.png "Иерархия исключений")

- **Проверяемые** - Исключения для которых компилятор будет требовать добавление try/catch или 
проброс вверх с помощью throws
- **Непроверяемые** - Данный тип исключений не обязательно обрабатывать.

## Управление ресурсами с помощью try (>= JDK 7 )

```java
try (java.io.AutoCloseabe resource1 = new SomeStream(); java.io.AutoCloseabe resource2 = new SomeStream();) {
   
} 
```

**Заметки**
* Что бы использовтаь ресурс в блоке try - он должен реализовывать интерфейс `java.io.AutoCloseabe`
